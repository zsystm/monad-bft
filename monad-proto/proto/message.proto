syntax = "proto3";

package monad_proto.message;

import "basic.proto";
import "block.proto";
import "blocksync.proto";
import "discovery.proto";
import "signing.proto";
import "timeout.proto";
import "voting.proto";


message ProtoVoteMessage {
  monad_proto.voting.ProtoVote vote = 1;
  monad_proto.signing.ProtoSignature sig = 2;
}

message ProtoBlockSyncRequestMessage {
  oneof request_type {
    monad_proto.blocksync.ProtoBlockRange block_range = 1;
    monad_proto.basic.ProtoBlockBodyId block_body_id = 2;
  }
}

message ProtoBlockSyncResponseMessage {
  oneof blocksync_response {
    monad_proto.blocksync.ProtoBlockSyncHeadersResponse headers_response = 1;
    monad_proto.blocksync.ProtoBlockSyncBodyResponse body_response = 2;
  }
}

message ProtoTimeoutMessage {
  monad_proto.timeout.ProtoTimeout timeout = 1;
  monad_proto.signing.ProtoSignature sig = 2;
}

message ProtoProposalMessage {
  monad_proto.block.ProtoBlockHeader block_header = 1;
  monad_proto.block.ProtoBlockBody block_body = 2;
  optional monad_proto.timeout.ProtoTimeoutCertificate last_round_tc = 3;
}

message ProtoUnverifiedConsensusMessage {
  oneof OneofMessage {
    ProtoProposalMessage proposal = 1;
    ProtoTimeoutMessage timeout = 2;
    ProtoVoteMessage vote = 3;
  }
  uint32 version = 4;
  monad_proto.signing.ProtoSignature author_signature = 6;
}

message ProtoForwardedTx {
  repeated bytes tx = 1;
}

message ProtoStateSyncRequest {
  uint32 version = 7;

  uint64 prefix = 1;
  uint32 prefix_bytes = 2;
  uint64 target = 3;
  uint64 from = 4;
  uint64 until = 5;
  uint64 old_target = 6;
}

enum ProtoStateSyncUpsertType {
  CODE = 0;
  ACCOUNT = 1;
  STORAGE = 2;
  ACCOUNT_DELETE = 3;
  STORAGE_DELETE = 4;
  HEADER = 5;
}

message ProtoStateSyncUpsert {
  ProtoStateSyncUpsertType type = 1;
  bytes data = 2;
}

message ProtoStateSyncResponse {
  uint32 version = 4;
  uint64 nonce = 5;
  uint32 response_index = 6;

  ProtoStateSyncRequest request = 1;
  repeated ProtoStateSyncUpsert upserts = 2;
  uint64 n = 3;
}

message ProtoStateSyncBadVersion {
  uint32 min_version = 1;
  uint32 max_version = 2;
}

message ProtoStateSyncCompletion {
  uint64 session_id = 1;
}

message ProtoStateSyncNetworkMessage {
  oneof OneofMessage {
    ProtoStateSyncRequest request = 1;
    ProtoStateSyncResponse response = 2;
    ProtoStateSyncBadVersion bad_version = 3;
    ProtoStateSyncCompletion completion = 4;
  }
}

message ProtoMonadMessage {
  oneof OneofMessage {
    ProtoUnverifiedConsensusMessage consensus = 1;
    ProtoBlockSyncRequestMessage block_sync_request = 2;
    ProtoBlockSyncResponseMessage block_sync_response = 3;
    ProtoForwardedTx forwarded_tx = 5;
    ProtoStateSyncNetworkMessage state_sync_message = 6;
    ProtoPingRequest ping_request = 7;
    ProtoPingResponse ping_response = 8;
  }
}
message ProtoDiscoveryRequest {
  monad_proto.discovery.ProtoMonadNameRecord self = 1;
}

message ProtoDiscoveryResponse {
  repeated monad_proto.discovery.ProtoMonadNameRecord peers = 1;
}

message ProtoDiscoveryMessage {
  oneof message {
    ProtoDiscoveryRequest request = 1;
    ProtoDiscoveryResponse response = 2;
  }
}

message ProtoRouterMessage {
  oneof message {
    // keep compatible with ProtoMonadMessage
    bytes app_message = 7;
    ProtoDiscoveryMessage discovery_message = 8;
  }
}

message ProtoPingRequest {
  uint64 sequence = 1;
}

message ProtoPingResponse {
  uint64 sequence = 1;
}