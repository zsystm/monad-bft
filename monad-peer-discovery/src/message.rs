// Copyright (C) 2025 Category Labs, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use alloy_rlp::{Decodable, Encodable, Header, RlpDecodable, RlpEncodable, encode_list};
use monad_crypto::certificate_signature::{
    CertificateSignaturePubKey, CertificateSignatureRecoverable,
};
use monad_executor_glue::Message;
use monad_types::NodeId;

use crate::{MonadNameRecord, PeerDiscoveryEvent};

const PEER_DISCOVERY_VERSION: u16 = 1;

#[derive(Debug, Clone)]
pub enum PeerDiscoveryMessage<ST: CertificateSignatureRecoverable> {
    Ping(Ping<ST>),
    Pong(Pong),
    PeerLookupRequest(PeerLookupRequest<ST>),
    PeerLookupResponse(PeerLookupResponse<ST>),
}

impl<ST: CertificateSignatureRecoverable> Message for PeerDiscoveryMessage<ST> {
    type NodeIdPubKey = CertificateSignaturePubKey<ST>;
    type Event = PeerDiscoveryEvent<ST>;

    fn event(self, from: NodeId<Self::NodeIdPubKey>) -> Self::Event {
        match self {
            PeerDiscoveryMessage::Ping(ping) => PeerDiscoveryEvent::PingRequest { from, ping },
            PeerDiscoveryMessage::Pong(pong) => PeerDiscoveryEvent::PongResponse { from, pong },
            PeerDiscoveryMessage::PeerLookupRequest(request) => {
                PeerDiscoveryEvent::PeerLookupRequest { from, request }
            }
            PeerDiscoveryMessage::PeerLookupResponse(response) => {
                PeerDiscoveryEvent::PeerLookupResponse { from, response }
            }
        }
    }
}

impl<ST: CertificateSignatureRecoverable> Encodable for PeerDiscoveryMessage<ST> {
    fn encode(&self, out: &mut dyn alloy_rlp::BufMut) {
        let version = PEER_DISCOVERY_VERSION;

        match self {
            PeerDiscoveryMessage::Ping(ping) => {
                let enc: [&dyn Encodable; 3] = [&version, &1_u8, ping];
                encode_list::<_, dyn Encodable>(&enc, out);
            }
            PeerDiscoveryMessage::Pong(pong) => {
                let enc: [&dyn Encodable; 3] = [&version, &2_u8, pong];
                encode_list::<_, dyn Encodable>(&enc, out);
            }
            PeerDiscoveryMessage::PeerLookupRequest(peer_lookup_request) => {
                let enc: [&dyn Encodable; 3] = [&version, &3_u8, peer_lookup_request];
                encode_list::<_, dyn Encodable>(&enc, out);
            }
            PeerDiscoveryMessage::PeerLookupResponse(peer_lookup_response) => {
                let enc: [&dyn Encodable; 3] = [&version, &4_u8, peer_lookup_response];
                encode_list::<_, dyn Encodable>(&enc, out);
            }
        }
    }
}

impl<ST: CertificateSignatureRecoverable> Decodable for PeerDiscoveryMessage<ST> {
    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
        let mut payload = Header::decode_bytes(buf, true)?;
        let _peer_discovery_version = u16::decode(&mut payload)?;

        match u8::decode(&mut payload)? {
            1 => Ok(Self::Ping(Ping::decode(&mut payload)?)),
            2 => Ok(Self::Pong(Pong::decode(&mut payload)?)),
            3 => Ok(Self::PeerLookupRequest(PeerLookupRequest::decode(
                &mut payload,
            )?)),
            4 => Ok(Self::PeerLookupResponse(PeerLookupResponse::decode(
                &mut payload,
            )?)),
            _ => Err(alloy_rlp::Error::Custom(
                "failed to decode unknown PeerDiscoveryMessage",
            )),
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, RlpDecodable, RlpEncodable)]
#[rlp(trailing)]
pub struct Ping<ST: CertificateSignatureRecoverable> {
    pub id: u32,
    pub local_name_record: Option<MonadNameRecord<ST>>,
}

impl<ST: CertificateSignatureRecoverable> Eq for Ping<ST> {}

#[derive(Debug, Clone, Copy, PartialEq, Eq, RlpDecodable, RlpEncodable)]
pub struct Pong {
    pub ping_id: u32,
    pub local_record_seq: u64,
}

#[derive(Debug, Clone, RlpDecodable, RlpEncodable)]
pub struct PeerLookupRequest<ST: CertificateSignatureRecoverable> {
    pub lookup_id: u32,
    pub target: NodeId<CertificateSignaturePubKey<ST>>,
    pub open_discovery: bool,
}

#[derive(Debug, Clone, RlpDecodable, RlpEncodable)]
pub struct PeerLookupResponse<ST: CertificateSignatureRecoverable> {
    pub lookup_id: u32,
    pub target: NodeId<CertificateSignaturePubKey<ST>>,
    pub name_records: Vec<MonadNameRecord<ST>>,
}

#[cfg(test)]
mod test {
    use std::{net::SocketAddrV4, str::FromStr};

    use monad_secp::SecpSignature;
    use monad_testutil::signing::get_key;

    use super::*;
    use crate::NameRecord;

    type SignatureType = SecpSignature;

    #[test]
    fn test_ping_with_record_rlp_roundtrip() {
        let key = get_key::<SignatureType>(37);
        let ping = Ping {
            id: 257,
            local_name_record: Some(MonadNameRecord::<SignatureType>::new(
                NameRecord {
                    ip: SocketAddrV4::from_str("127.0.0.1:8000")
                        .unwrap()
                        .ip()
                        .clone(),
                    tcp_port: 8000,
                    udp_port: 8000,
                    direct_udp_port: None,
                    capabilities: 0,
                    seq: 2,
                },
                &key,
            )),
        };

        let mut encoded = Vec::new();
        ping.encode(&mut encoded);

        let decoded = Ping::<SignatureType>::decode(&mut encoded.as_slice()).unwrap();
        assert_eq!(ping, decoded);
    }

    #[test]
    fn test_ping_without_record_rlp_roundtrip() {
        let ping = Ping {
            id: 257,
            local_name_record: None,
        };

        let mut encoded = Vec::new();
        ping.encode(&mut encoded);

        let decoded = Ping::<SignatureType>::decode(&mut encoded.as_slice()).unwrap();
        assert_eq!(ping, decoded);
    }
}
