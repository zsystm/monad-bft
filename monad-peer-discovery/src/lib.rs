// Copyright (C) 2025 Category Labs, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use std::{
    collections::{BTreeSet, HashMap},
    net::{Ipv4Addr, SocketAddrV4},
    time::Duration,
};

use alloy_rlp::{Decodable, Encodable, RlpDecodable, RlpEncodable, encode_list};
use message::{PeerLookupRequest, PeerLookupResponse, Ping, Pong};
use monad_crypto::{
    certificate_signature::{
        CertificateSignature, CertificateSignaturePubKey, CertificateSignatureRecoverable,
    },
    signing_domain,
};
use monad_executor::ExecutorMetrics;
use monad_executor_glue::PeerEntry;
use monad_types::{Epoch, NodeId, Round};
use smallvec::SmallVec;
use tracing::warn;

pub mod discovery;
pub mod driver;
pub mod message;
pub mod mock;

pub use message::PeerDiscoveryMessage;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum PortTag {
    TCP = 0,
    UDP = 1,
    DirectUdp = 2,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, RlpEncodable, RlpDecodable)]
pub struct Port {
    pub tag: u8,
    pub port: u16,
}

impl Port {
    pub fn new(tag: PortTag, port: u16) -> Self {
        Self {
            tag: tag as u8,
            port,
        }
    }

    pub fn tag_enum(&self) -> Option<PortTag> {
        match self.tag {
            0 => Some(PortTag::TCP),
            1 => Some(PortTag::UDP),
            2 => Some(PortTag::DirectUdp),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct WireNameRecord {
    pub ip: Ipv4Addr,
    pub ports: SmallVec<[Port; 3]>,
    pub capabilities: u64,
    pub seq: u64,
}

impl Encodable for WireNameRecord {
    fn encode(&self, out: &mut dyn alloy_rlp::BufMut) {
        let ports_vec: Vec<Port> = self.ports.to_vec();
        let enc: [&dyn Encodable; 4] =
            [&self.ip.octets(), &ports_vec, &self.capabilities, &self.seq];
        encode_list::<_, dyn Encodable>(&enc, out);
    }
}

fn decode_smallvec_with_limit<T: Decodable, const N: usize>(
    buf: &mut &[u8],
    limit: usize,
) -> alloy_rlp::Result<SmallVec<[T; N]>> {
    let mut bytes = alloy_rlp::Header::decode_bytes(buf, true)?;
    let mut vec = SmallVec::new();
    let payload_view = &mut bytes;
    while !payload_view.is_empty() {
        if vec.len() >= limit {
            return Err(alloy_rlp::Error::Custom("Too many items in vector"));
        }
        vec.push(T::decode(payload_view)?);
    }
    Ok(vec)
}

impl Decodable for WireNameRecord {
    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
        let buf = &mut alloy_rlp::Header::decode_bytes(buf, true)?;

        let Ok(ip_bytes) = <[u8; 4]>::decode(buf) else {
            warn!("ip address decode failed: {:?}", buf);
            return Err(alloy_rlp::Error::Custom("Invalid IPv4 address"));
        };
        let ip = Ipv4Addr::from(ip_bytes);
        let ports = decode_smallvec_with_limit::<Port, 3>(buf, 8)?;
        let capabilities = u64::decode(buf)?;
        let seq = u64::decode(buf)?;

        Ok(Self {
            ip,
            ports,
            capabilities,
            seq,
        })
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct NameRecord {
    pub ip: Ipv4Addr,
    pub tcp_port: u16,
    pub udp_port: u16,
    pub direct_udp_port: Option<u16>,
    pub capabilities: u64,
    pub seq: u64,
}

impl Encodable for NameRecord {
    fn encode(&self, out: &mut dyn alloy_rlp::BufMut) {
        let mut ports = SmallVec::new();
        ports.push(Port::new(PortTag::TCP, self.tcp_port));
        ports.push(Port::new(PortTag::UDP, self.udp_port));

        if let Some(direct_udp_port) = self.direct_udp_port {
            ports.push(Port::new(PortTag::DirectUdp, direct_udp_port));
        }

        let wire = WireNameRecord {
            ip: self.ip,
            ports,
            capabilities: self.capabilities,
            seq: self.seq,
        };

        wire.encode(out);
    }
}

impl Decodable for NameRecord {
    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
        let wire = WireNameRecord::decode(buf)?;

        let mut tcp_port = None;
        let mut udp_port = None;
        let mut direct_udp_port = None;
        let mut seen_tags = std::collections::HashSet::new();

        for port in &wire.ports {
            if !seen_tags.insert(port.tag) {
                return Err(alloy_rlp::Error::Custom("duplicate port tag"));
            }

            match port.tag_enum() {
                Some(PortTag::TCP) => tcp_port = Some(port.port),
                Some(PortTag::UDP) => udp_port = Some(port.port),
                Some(PortTag::DirectUdp) => direct_udp_port = Some(port.port),
                None => return Err(alloy_rlp::Error::Custom("Invalid port tag")),
            }
        }

        let tcp_port = tcp_port.ok_or(alloy_rlp::Error::Custom("Missing TCP port"))?;
        let udp_port = udp_port.ok_or(alloy_rlp::Error::Custom("Missing UDP port"))?;

        Ok(NameRecord {
            ip: wire.ip,
            tcp_port,
            udp_port,
            direct_udp_port,
            capabilities: wire.capabilities,
            seq: wire.seq,
        })
    }
}

impl NameRecord {
    pub fn tcp_socket(&self) -> SocketAddrV4 {
        SocketAddrV4::new(self.ip, self.tcp_port)
    }

    pub fn udp_socket(&self) -> SocketAddrV4 {
        SocketAddrV4::new(self.ip, self.udp_port)
    }

    pub fn direct_udp_socket(&self) -> Option<SocketAddrV4> {
        self.direct_udp_port
            .map(|port| SocketAddrV4::new(self.ip, port))
    }

    pub fn check_capability(&self, capability: Capability) -> bool {
        (self.capabilities & (1u64 << (capability as u8))) != 0
    }

    pub fn set_capability(&mut self, capability: Capability) {
        self.capabilities |= 1u64 << (capability as u8);
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Capability {}

#[derive(Debug, Clone, Copy, PartialEq, RlpEncodable, RlpDecodable, Eq)]
pub struct MonadNameRecord<ST: CertificateSignatureRecoverable> {
    pub name_record: NameRecord,
    pub signature: ST,
}

impl<ST: CertificateSignatureRecoverable + Copy> MonadNameRecord<ST> {
    pub fn new(name_record: NameRecord, key: &ST::KeyPairType) -> Self {
        let mut encoded = Vec::new();
        name_record.encode(&mut encoded);
        let signature = ST::sign::<signing_domain::NameRecord>(&encoded, key);
        Self {
            name_record,
            signature,
        }
    }

    pub fn recover_pubkey(
        &self,
    ) -> Result<NodeId<CertificateSignaturePubKey<ST>>, <ST as CertificateSignature>::Error> {
        let mut encoded = Vec::new();
        self.name_record.encode(&mut encoded);
        let pubkey = self
            .signature
            .recover_pubkey::<signing_domain::NameRecord>(&encoded)?;
        Ok(NodeId::new(pubkey))
    }

    pub fn address(&self) -> SocketAddrV4 {
        self.name_record.tcp_socket()
    }

    pub fn seq(&self) -> u64 {
        self.name_record.seq
    }
}

#[derive(Debug, Clone)]
pub enum PeerDiscoveryEvent<ST: CertificateSignatureRecoverable> {
    SendPing {
        to: NodeId<CertificateSignaturePubKey<ST>>,
    },
    PingRequest {
        from: NodeId<CertificateSignaturePubKey<ST>>,
        ping: Ping<ST>,
    },
    PongResponse {
        from: NodeId<CertificateSignaturePubKey<ST>>,
        pong: Pong,
    },
    PingTimeout {
        to: NodeId<CertificateSignaturePubKey<ST>>,
        ping_id: u32,
    },
    SendPeerLookup {
        to: NodeId<CertificateSignaturePubKey<ST>>,
        target: NodeId<CertificateSignaturePubKey<ST>>,
        open_discovery: bool,
    },
    PeerLookupRequest {
        from: NodeId<CertificateSignaturePubKey<ST>>,
        request: PeerLookupRequest<ST>,
    },
    PeerLookupResponse {
        from: NodeId<CertificateSignaturePubKey<ST>>,
        response: PeerLookupResponse<ST>,
    },
    PeerLookupTimeout {
        to: NodeId<CertificateSignaturePubKey<ST>>,
        target: NodeId<CertificateSignaturePubKey<ST>>,
        lookup_id: u32,
    },
    UpdateCurrentRound {
        round: Round,
        epoch: Epoch,
    },
    UpdateValidatorSet {
        epoch: Epoch,
        validators: BTreeSet<NodeId<CertificateSignaturePubKey<ST>>>,
    },
    UpdatePeers {
        peers: Vec<PeerEntry<ST>>,
    },
    UpdateConfirmGroup {
        end_round: Round,
        peers: BTreeSet<NodeId<CertificateSignaturePubKey<ST>>>,
    },
    Refresh,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum TimerKind {
    SendPing,
    PingTimeout,
    RetryPeerLookup { lookup_id: u32 },
    Refresh,
}

#[derive(Debug)]
pub enum PeerDiscoveryTimerCommand<E, ST: CertificateSignatureRecoverable> {
    Schedule {
        node_id: NodeId<CertificateSignaturePubKey<ST>>,
        timer_kind: TimerKind,
        duration: Duration,
        on_timeout: E,
    },
    ScheduleReset {
        node_id: NodeId<CertificateSignaturePubKey<ST>>,
        timer_kind: TimerKind,
    },
}

#[derive(Debug)]
pub struct PeerDiscoveryMetricsCommand(ExecutorMetrics);

#[derive(Debug)]
pub enum PeerDiscoveryCommand<ST: CertificateSignatureRecoverable> {
    RouterCommand {
        target: NodeId<CertificateSignaturePubKey<ST>>,
        message: PeerDiscoveryMessage<ST>,
    },
    TimerCommand(PeerDiscoveryTimerCommand<PeerDiscoveryEvent<ST>, ST>),
    MetricsCommand(PeerDiscoveryMetricsCommand),
}

pub trait PeerDiscoveryAlgo {
    type SignatureType: CertificateSignatureRecoverable;

    fn send_ping(
        &mut self,
        target: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_ping(
        &mut self,
        from: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        ping: Ping<Self::SignatureType>,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_pong(
        &mut self,
        from: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        pong: Pong,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_ping_timeout(
        &mut self,
        to: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        ping_id: u32,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn send_peer_lookup_request(
        &mut self,
        to: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        target: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        open_discovery: bool,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_peer_lookup_request(
        &mut self,
        from: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        request: PeerLookupRequest<Self::SignatureType>,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_peer_lookup_response(
        &mut self,
        from: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        response: PeerLookupResponse<Self::SignatureType>,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn handle_peer_lookup_timeout(
        &mut self,
        to: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        target: NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        lookup_id: u32,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn refresh(&mut self) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn update_current_round(
        &mut self,
        round: Round,
        epoch: Epoch,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn update_validator_set(
        &mut self,
        epoch: Epoch,
        validators: BTreeSet<NodeId<CertificateSignaturePubKey<Self::SignatureType>>>,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn update_peers(
        &mut self,
        peers: Vec<PeerEntry<Self::SignatureType>>,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn update_peer_participation(
        &mut self,
        round: Round,
        peers: BTreeSet<NodeId<CertificateSignaturePubKey<Self::SignatureType>>>,
    ) -> Vec<PeerDiscoveryCommand<Self::SignatureType>>;

    fn metrics(&self) -> &ExecutorMetrics;

    fn get_addr_by_id(
        &self,
        id: &NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
    ) -> Option<SocketAddrV4>;

    fn get_known_addrs(
        &self,
    ) -> HashMap<NodeId<CertificateSignaturePubKey<Self::SignatureType>>, SocketAddrV4>;

    fn get_name_records(
        &self,
    ) -> HashMap<
        NodeId<CertificateSignaturePubKey<Self::SignatureType>>,
        MonadNameRecord<Self::SignatureType>,
    >;
}

pub trait PeerDiscoveryAlgoBuilder {
    type PeerDiscoveryAlgoType: PeerDiscoveryAlgo;

    fn build(
        self,
    ) -> (
        Self::PeerDiscoveryAlgoType,
        Vec<
            PeerDiscoveryCommand<<Self::PeerDiscoveryAlgoType as PeerDiscoveryAlgo>::SignatureType>,
        >,
    );
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use super::*;

    #[test]
    fn test_name_record_v4_rlp() {
        let name_record = NameRecord {
            ip: Ipv4Addr::from_str("1.1.1.1").unwrap(),
            tcp_port: 8000,
            udp_port: 8001,
            direct_udp_port: None,
            capabilities: 0,
            seq: 2,
        };

        let mut encoded = Vec::new();
        name_record.encode(&mut encoded);

        let result = NameRecord::decode(&mut encoded.as_slice());
        assert!(result.is_ok());
        assert_eq!(result.unwrap(), name_record);
    }

    #[test]
    fn test_name_record_validation() {
        let name_record = NameRecord {
            ip: Ipv4Addr::from_str("1.1.1.1").unwrap(),
            tcp_port: 8000,
            udp_port: 8001,
            direct_udp_port: None,
            capabilities: 0,
            seq: 1,
        };

        assert_eq!(
            name_record.tcp_socket(),
            SocketAddrV4::from_str("1.1.1.1:8000").unwrap()
        );
        assert_eq!(
            name_record.udp_socket(),
            SocketAddrV4::from_str("1.1.1.1:8001").unwrap()
        );
        assert_eq!(name_record.direct_udp_socket(), None);

        let name_record_with_direct = NameRecord {
            ip: Ipv4Addr::from_str("1.1.1.1").unwrap(),
            tcp_port: 8000,
            udp_port: 8001,
            direct_udp_port: Some(8002),
            capabilities: 0,
            seq: 1,
        };
        assert_eq!(
            name_record_with_direct.direct_udp_socket(),
            Some(SocketAddrV4::from_str("1.1.1.1:8002").unwrap())
        );
    }

    #[test]
    fn test_name_record_duplicate_port_validation() {
        let wire = WireNameRecord {
            ip: Ipv4Addr::from_str("1.1.1.1").unwrap(),
            ports: smallvec::smallvec![
                Port::new(PortTag::TCP, 8000),
                Port::new(PortTag::UDP, 8001),
                Port::new(PortTag::TCP, 8002),
            ],
            capabilities: 0,
            seq: 1,
        };

        let mut encoded = Vec::new();
        wire.encode(&mut encoded);

        let decoded = NameRecord::decode(&mut encoded.as_slice());
        assert!(decoded.is_err());
    }

    #[test]
    fn test_name_record_missing_port_validation() {
        let wire = WireNameRecord {
            ip: Ipv4Addr::from_str("1.1.1.1").unwrap(),
            ports: smallvec::smallvec![Port::new(PortTag::TCP, 8000)],
            capabilities: 0,
            seq: 1,
        };

        let mut encoded = Vec::new();
        wire.encode(&mut encoded);

        let decoded = NameRecord::decode(&mut encoded.as_slice());
        assert!(decoded.is_err());
    }

    #[test]
    fn test_name_record_encode_snapshot() {
        let name_record = NameRecord {
            ip: Ipv4Addr::from_str("192.168.1.1").unwrap(),
            tcp_port: 8080,
            udp_port: 8081,
            direct_udp_port: None,
            capabilities: 0,
            seq: 42,
        };

        let mut encoded = Vec::new();
        name_record.encode(&mut encoded);
        let hex_encoded = hex::encode(&encoded);
        insta::assert_snapshot!(hex_encoded);
    }

    #[test]
    fn test_name_record_with_direct_udp_encode_snapshot() {
        let name_record = NameRecord {
            ip: Ipv4Addr::from_str("10.0.0.1").unwrap(),
            tcp_port: 9000,
            udp_port: 9001,
            direct_udp_port: Some(9002),
            capabilities: 1,
            seq: 100,
        };

        let mut encoded = Vec::new();
        name_record.encode(&mut encoded);
        let hex_encoded = hex::encode(&encoded);
        insta::assert_snapshot!(hex_encoded);
    }

    #[test]
    fn test_name_record_roundtrip() {
        let original = NameRecord {
            ip: Ipv4Addr::from_str("172.16.0.1").unwrap(),
            tcp_port: 7000,
            udp_port: 7001,
            direct_udp_port: Some(7002),
            capabilities: 255,
            seq: 999,
        };

        let mut encoded = Vec::new();
        original.encode(&mut encoded);

        let decoded = NameRecord::decode(&mut encoded.as_slice()).unwrap();
        assert_eq!(original, decoded);
    }

    #[test]
    fn test_wire_name_record_compatibility() {
        let name_record = NameRecord {
            ip: Ipv4Addr::from_str("127.0.0.1").unwrap(),
            tcp_port: 8000,
            udp_port: 8001,
            direct_udp_port: Some(8002),
            capabilities: 0,
            seq: 1,
        };

        let mut encoded = Vec::new();
        name_record.encode(&mut encoded);

        let wire = WireNameRecord::decode(&mut encoded.as_slice()).unwrap();
        assert_eq!(wire.ports.len(), 3);
        assert_eq!(wire.ports[0].tag, PortTag::TCP as u8);
        assert_eq!(wire.ports[0].port, 8000);
        assert_eq!(wire.ports[1].tag, PortTag::UDP as u8);
        assert_eq!(wire.ports[1].port, 8001);
        assert_eq!(wire.ports[2].tag, PortTag::DirectUdp as u8);
        assert_eq!(wire.ports[2].port, 8002);
    }
}
