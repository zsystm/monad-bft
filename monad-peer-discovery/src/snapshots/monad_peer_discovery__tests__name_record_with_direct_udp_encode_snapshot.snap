---
source: monad-peer-discovery/src/lib.rs
assertion_line: 582
expression: hex_encoded
---
d7840a000001cfc480822328c401822329c40282232a0164
